Alex and Happy Part 1/article.md

```diff
@@ -0,0 +1,728 @@
+Like Haskell, MiniML will support passing operators as first-class functions, like `let add = (+)`.
+However, this has the same deficiency as OCaml: `(*)` is recognized as a comment.
+However, `( *)` is accepted by OCaml but emits a message saying, "Warning : this is not the end of a comment.".
+Because of this, we will emit an error using `alexError` to the user and expect that `( * )` will be used to reference the multiplication operator instead.
+Alternatively, you can change comments' syntax to something else here, such as `/* */`, and avoid this ambiguity, but that wouldn't feel too ML-like.
+
+Next, we will change `AlexUserState` so it remembers the nesting level of our comments. We will initialize it with 0.
+We also define some helper functions for dealing with our state.
+
+```haskell
+data AlexUserState = AlexUserState
+  { nestLevel :: Int
+  }
+
+alexInitUserState :: AlexUserState
+alexInitUserState = AlexUserState
+  { nestLevel = 0
+  }
+
+get :: Alex AlexUserState
+get = Alex $ \s -> Right (s, alex_ust s)
+
+set :: AlexUserState -> Alex ()
+set s' = Alex $ \s -> Right (s{alex_ust = s'}, ())
+
+modify :: (AlexUserState -> AlexUserState) -> Alex ()
+modify f = Alex $ \s -> Right (s{alex_ust = f (alex_ust s)}, ())
+```
+
+We simply increase the nesting value and skip the consumed input to nest a comment.
+Note that `skip input len` is simply `alexMonadScan`, which asks to scan the next token.
+Unnesting is similar, except that we decrease the level, and if we reach level 0, we must return to our initial start code (`0`) as it means we exited the comment.
+Don't forget to import `when` from `Control.Monad`.
+
+```haskell
+nestComment, unnestComment :: AlexAction Token
+nestComment input len = do
+  modify $ \s -> s{nestLevel = nestLevel s + 1}
+  skip input len
+unnestComment input len = do
+  state <- get
+  let level = nestLevel state - 1
+  set state{nestLevel = level}
+  when (level == 0) $
+    alexSetStartCode 0
+  skip input len
+```
+
+Finally, let's change `alexEOF` to emit an error if we end up with an unclosed comment by checking whether we were parsing a comment when we reached EOF.
+
+```haskell
+alexEOF :: Alex Token
+alexEOF = do
+  startCode <- alexGetStartCode
+  when (startCode == comment) $
+    alexError "Error: unclosed comment"
+  (pos, _, _, _) <- alexGetInput
+  pure $ Token EOF (Range pos pos)
+```
+
+Trying it out in GHCi yields the expected results.
+
+```haskell
+>>> scanMany "(* my comment"
+Left "Error: unclosed comment"
+>>> scanMany "if (* my (* nested *) \n comment *) x"
+Right
+  [ Token {tType = If, tRange = Range {start = AlexPn 0 1 1, stop = AlexPn 2 1 3}}
+  , Token {tType = Identifier "x", tRange = Range {start = AlexPn 35 2 13, stop = AlexPn 36 2 14}}
+  , Token {tType = EOF, tRange = Range {start = AlexPn 36 2 14, stop = AlexPn 36 2 14}}
+  ]
+```
+
+#### Lexing strings
+
+Lexing strings should have no surprises. Just stick this in your lexer:
+
+```alex
+<0> \".*\"  { tokString }
```

Comment:
Oh, no. This is pretty bad, actually. Consider that this matches `"string" id "other string"`. At least use `\"[^\"]*\"`.

Reply:
Oh, that's a good point! I changed it.